src/Compilers/CSharp/Portable/Binder/Binder_AnonymousTypes.cs

```diff
@@ -203,13 +205,14 @@ private bool IsAnonymousTypesAllowed()
         /// </summary>
         private TypeSymbol GetAnonymousTypeFieldType(BoundExpression expression, CSharpSyntaxNode errorSyntax, DiagnosticBag diagnostics, ref bool hasError)
         {
-            object errorArg = null;
-            TypeSymbol expressionType = expression.Type;
+            object? errorArg = null;
+            TypeSymbol? expressionType = expression.Type;
 
             if (!expression.HasAnyErrors)
             {
                 if (expression.HasExpressionType())
                 {
+                    RoslynDebug.Assert(expressionType is object);
```

Comment:
Consider `!` here given that `HasExpressionType` has already asserted this.

Reply:
[https://github.com/dotnet/roslyn/issues/40992](https://github.com/dotnet/roslyn/issues/40992). It would require a lot of `!`, so I'm going to keep this as is.

---
In reply to: [367115054](https://github.com/dotnet/roslyn/pull/40946#discussion_r367115054) [](ancestors = 367115054)